app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,481 @@
+import androidx.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.util.ArrayList;
+import javax.inject.Inject;
+import kotlin.Unit;
+import kotlin.jvm.functions.Function0;
+import org.kiwix.kiwixmobile.BuildConfig;
+import org.kiwix.kiwixmobile.R;
+import org.kiwix.kiwixmobile.utils.AlertDialogShower;
+import org.kiwix.kiwixmobile.utils.KiwixDialog;
+import org.kiwix.kiwixmobile.utils.SharedPreferenceUtil;
+
+import static android.os.Looper.getMainLooper;
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.FileItem.FileStatus.ERROR;
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity.showToast;
+
+
+/**
+ * Manager for the Wifi-P2p API, used in the local file transfer module
+ * */
+public class WifiDirectManager implements WifiP2pManager.ChannelListener, WifiP2pManager.PeerListListener, WifiP2pManager.ConnectionInfoListener,
+    KiwixWifiP2pBroadcastReceiver.P2pEventListener {
+
+  private static final String TAG = "WifiDirectManager";
+  public static int FILE_TRANSFER_PORT = 8008;
+
+  private @NonNull LocalFileTransferActivity activity;
+  private @NonNull Callbacks callbacks;
+
+  private SharedPreferenceUtil sharedPreferenceUtil;
+  private AlertDialogShower alertDialogShower;
+
+  /* Variables related to the WiFi P2P API */
+  private boolean isWifiP2pEnabled = false; // Whether WiFi has been enabled or not
+  private boolean retryChannel = false;   // Whether channel has retried connecting previously
+
+  private WifiP2pManager manager;         // Overall manager of Wifi p2p connections for the module
+  private WifiP2pManager.Channel channel;
+  // Connects the module to device's underlying Wifi p2p framework
+
+  private BroadcastReceiver receiver = null; // For receiving the broadcasts given by above filter
+
+  private WifiP2pInfo groupInfo;      // Corresponds to P2P group formed between the two devices
+
+  private WifiP2pDevice senderSelectedPeerDevice = null;
+
+  private PeerGroupHandshakeAsyncTask peerGroupHandshakeAsyncTask;
+  private SenderDeviceAsyncTask senderDeviceAsyncTask;
+  private ReceiverDeviceAsyncTask receiverDeviceAsyncTask;
+
+  private InetAddress selectedPeerDeviceInetAddress;
+  private InetAddress fileReceiverDeviceAddress;  // IP address of the file receiving device
+
+  private int totalFilesForTransfer = -1;
+  private int filesSent = 0;          // Count of number of files transferred until now
+  private ArrayList<FileItem> filesForTransfer;
+
+  private boolean isFileSender = false;    // Whether the device is the file sender or not
+  private boolean hasSenderStartedConnection = false;
+
+  @Inject
+  public WifiDirectManager(@NonNull Activity activity, @NonNull SharedPreferenceUtil sharedPreferenceUtil,
+      @NonNull AlertDialogShower alertDialogShower) {
+    this.activity = (LocalFileTransferActivity) activity;
```

Comment:
this should not cast to LFTA,  WDM should know nothing about where it is used so the type of `activity` should be Activity